### ISSUE TITLE
[messaging] Ability to change the notification before it's sent to the system tray

### ISSUE THREAD

**Issue by helenaford** [firebase/firebase-android-sdk#2639]
## What feature would you like to see?

Hi, is there anything on the roadmap to provide the ability to customise the notification before it's sent to the system tray.  I'm a maintainer of [notifee](https://github.com/notifee/react-native-notifee), and we recently added support to do this with iOS, we'd love to be able to do it with Android too.

Currently, the only option there seems to be it is to send data messages. A couple of issues with this method are that you need to send two different types of messages based on if the device is android or iOS, and this also requires to keep track of which the platform the token was registered with. 

In addition to platform-dependent code, there's also different background restrictions for data-only vs notification messages (ref [here](https://firebase.google.com/docs/cloud-messaging/concept-options#setting-the-priority-of-a-message)). 

On iOS, you can alter the notification using a Notification Service Extension where you have a brief amount of time to modify the notification before it's displayed to the user. As well as the `message.apns.aps` key offers a lot of flexibility where you can add a categoryId, sound, threadId e.t.c to the payload. 

One possible solution could be to allow [FirebaseMessagingService.onMessageReceived](https://firebase.google.com/docs/cloud-messaging/android/receive#handling_messages)  to be called for all app states, maybe a check somehow to see if the app does supports this, if not, default to handling with FCM? If this was possible, the one issue I do see with this, is that it could clash with other libraries like `FlutterFire` and `react-native-firebase` as you can only have one FirebaseMessagingService per app. 

Another solution, which I think seems way easier and less intrusive is to have a separate service, like how iOS works, to alter the notification and nothing else.

I can see how this could be a big f/r, and maybe there's reasons that it's not been done?

The other possible solution is to expand the predefined set of user-visible keys for `[message.android](https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#androidnotification)`, to include `group` and other `NotificationCompat` properties where possible. 

To summarise, the two pain points I think there is at the moment with Android is:
 - the options allowed via [message.android](https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#androidnotification) is limited.
 - no ability to intercept the notification before it's shown when the app is in background
 
## How would you use it?

Tell us how you'd use this feature in your app.

This feature would allow users of this library to send notification messages and customise the notification before it's finally sent to the system tray.

**Comment by sfuqua** [comment]
The comparison to the iOS Notification Service Extension is a good one - I think Apple has a fairly elegant solution here in allowing an app to effectively "tamper" with a notification payload before it's handled by the system. I'd love to do exactly the same thing in Android Firebase land - mutate the RemoteMessage before the SDK handles it.

I have legacy services that are sending a notification payload with my FCM messages. It's not impossible to change these to data only, but it's costly, as it's a breaking change for existing clients. Allowing a client to intercept the payload and remove the "notification" blob before the SDK handles it would be a godsend.

**Comment by sfuqua** [comment]
To spitball a surgical proposed change without more edits on my previous post, it'd basically be:

1. New overrideable method on `FirebaseMessagingService`:
```java
/**
 * Called when determining whether to opt an Intent out of default system tray behavior for a "notification message".
 * @return true if this message should fall-through to {@link #onMessageReceived}, false for default system behavior.
 */
public boolean shouldOptOutOfNotificationBehavior(@NonNull RemoteMessage message) { return false; }
```

2. Update https://github.com/firebase/firebase-android-sdk/blob/b2aed41639c6a9590fa4d31098805763ceb6eb9f/firebase-messaging/src/main/java/com/google/firebase/messaging/FirebaseMessagingService.java#L215 to call `this.shouldOptOutOfNotificationBehavior(new RemoteMessage(data))` as an `&&` clause with the existing `isNotification` check.

unless I am misunderstanding, this would unblock the desired behavior of allowing an app to configure the rendered notification while backgrounded, in an opt-out fashion so existing clients are unaffected by default.

I'd love to allow _modifying_ the Intent/Bundle before the SDK handles it, but since there seems to be a bunch of protected GCM keys stuffed into the payload I can see why we may not want to expose that to an application. However, allowing an app to inspect the corresponding `RemoteMessage` as a proxy and make a "go"/"no-go" decision for the system tray based on that payload is almost as good?

**Comment by gsakakihara** [comment]
Something similar has been discussed and proposed in the past internally and has come up again recently, but we don't have anything new to report at this time. Support for this is mostly a product question that comes down to prioritization, demand for the feature, and some other considerations.

**Comment by sfuqua** [comment]
For anyone following this issue - since the Firebase team is not showing any interest in my PR to tweak the API for this scenario, my app team has been working around this problem with a custom `FirebaseMessagingService` that overrides `handleIntent` to intercept specific notification Intents before they're sent to Firebase SDK and automatically displayed, and instead displaying them ourselves.

This _seems_ to work just fine despite being an undocumented workaround. I have not received any reports of it causing problems despite being live for a few months.

The approach is based on how the SDK's service works under the covers and is something like:

```java
// this is inside a custom FirebaseMessagingService which we've hooked up in our AndroidManifest

// 1. excessively document for your team with links to GitHub issues, PRs, StackOverflow, etc
// 2. emphasize that this is a hack which risks breaking in future Firebase SDK updates :(
// 3. dream of an officially supported API for this scenario
@Override
public void handleIntent(Intent intent) {
  Bundle extras = intent.getExtras();
  
  // try/catch around this to be defensive; default to super.handleIntent
  RemoteMessage message = new RemoteMessage(extras);
  
  RemoteMessage.Notification notificationData = message.getNotification();
  // short-circuit if this is not a display notification
  if (notificationData == null) { super.handleIntent(intent); return; }
  
  // 1. filter on the notification however you want using the data payload, clickAction, etc
  // 2. manually display notifications that match your filter, and then return
  // 3. fall-through to this for everything else...
  super.handleIntent(intent);
}
```

**Comment by andreibarabas** [comment]
thanks @sfuqua for suggesting this. your approach makes total sense. will use it a starting base

**Comment by 0x090909** [comment]
ping any news?

**Comment by 0x090909** [comment]
@sfuqua Thank you ! It works for me as well

**Comment by aureosouza** [comment]
@sfuqua thank you for this, we were able to implement similar approach overriding the handleIntent from the class that extends `FirebaseMessagingService` (in our case we are using react-native-firebase, so that would be ReactNativeFirebaseMessagingService). @helenaford We're using notifee as well, so we've added an extra fetchNotificationData to getDisplayedNotifications module method, that injects the data param from firebase (if it exists):

```java
private void createNotificationChannel() {
    // Same as Firebase SDK default channel name and ids
    NotificationChannel channel = new NotificationChannel("fcm_fallback_notification_channel", "Miscellaneous", NotificationManager.IMPORTANCE_HIGH);
    NotificationManager notificationManager = getSystemService(NotificationManager.class);
    notificationManager.createNotificationChannel(channel);
  }

  private int getNotificationIcon() {
    int iconResId;
    iconResId = getResources().getIdentifier("ic_notification", "drawable", getPackageName());
    if (iconResId == 0) {
      iconResId = getApplicationInfo().icon;
    }

    return iconResId;
  }

  @Override
  public void handleIntent(Intent intent) {
    Bundle extras = intent.getExtras();

    try {
      RemoteMessage message = new RemoteMessage(extras);
      RemoteMessage.Notification notificationData = message.getNotification();

      if (notificationData == null) {
        super.handleIntent(intent);
        return;
      }

      Bundle customExtras = new Bundle();
      for (Map.Entry<String, String> entry : message.getData().entrySet()) {
        customExtras.putString(entry.getKey(), entry.getValue());
      }
      createNotificationChannel();

      NotificationCompat.Builder notificationBuilder = new NotificationCompat.Builder(this, "fcm_fallback_notification_channel")
        .setSmallIcon(getNotificationIcon())
        .setContentTitle(notificationData.getTitle())
        .setContentText(notificationData.getBody())
        .setAutoCancel(true)
        .setPriority(NotificationCompat.PRIORITY_HIGH);

      notificationBuilder.getExtras().putBundle("data", customExtras);
      NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
      // Same as tag from Firebase SDK is built
      notificationManager.notify("FCM-Notification:" + SystemClock.uptimeMillis(), 0, notificationBuilder.build());

    } catch (Exception e) {
      super.handleIntent(intent);
    }
  }
```

And in Notifee module:
```java
private List<Bundle> fetchNotificationData(List<Bundle> aBundleList) {
    NotificationManager notificationManager = (NotificationManager) getReactApplicationContext().getSystemService(Context.NOTIFICATION_SERVICE);
    if (notificationManager != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
      StatusBarNotification[] activeNotifications = notificationManager.getActiveNotifications();

      for (StatusBarNotification sbn : activeNotifications) {
        Notification notification = sbn.getNotification();
        Bundle extras = notification.extras;

        if (extras != null) {
          Bundle data = extras.getBundle("data");
          if (data != null) {
            for (Bundle originalBundle : aBundleList) {
              Bundle originalNotificationBundle = originalBundle.getBundle("notification");
              if (originalNotificationBundle != null && originalNotificationBundle.getString("id").equals(String.valueOf(sbn.getId()))) {
                originalNotificationBundle.putBundle("data", data);
              }
            }
          }
        }
      }
    }
    return aBundleList;
  }


  @ReactMethod
  public void getDisplayedNotifications(Promise promise) {
    Notifee.getInstance()
      .getDisplayedNotifications(
        (e, aBundleList) -> NotifeeReactUtils.promiseResolver(promise, e, fetchNotificationData(aBundleList)));
  }
```
